modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java

```diff
@@ -94,7 +98,15 @@ protected void register(Context ctx, XAnnotatedObject xObject, Element element)
         Object contrib;
         XAnnotatedMember merge = xObject.getMerge();
         if (merge != null && Boolean.TRUE.equals(merge.getValue(ctx, element))) {
-            contrib = xObject.newInstance(ctx, element, contributions.get(id));
+            Object contribution = contributions.get(id);
+            if (contribution != null && xObject.getCompatWarnOnMerge() && !merge.hasValue(ctx, element)) {
+                log.warn(String.format(
+                        "The contribution with id '%s' has been implicitely merged: the compatibility mechanism on its"
+                                + " descriptor class '%s' detected it, and"
+                                + " the attribute merge=\"true\" should be added to this definition.",
+                        id, contribution.getClass().getName()));
```

Comment:
You should use `{}` parameters instead of `String.format` in logs. Same in `SingleRegistry`.